### PR TITLE
C backend fixes

### LINKED ISSUE #11651: C backend on behavior tests emits C code with compile errors when using -std=c99 -Werror -pedantic
### Zig Version

0.10.0-dev.2211+f32928c50

### Steps to Reproduce

```
stage2/bin/zig test ../test/behavior.zig -I../test -ofmt=c -femit-bin=test.c
stage2/bin/zig cc -c test.c -Werror -pedantic -std=c99
```

### Expected Behavior

No output, exit code 0.

### Actual Behavior

```
test.c:937:17: error: redefinition of typedef 'zig_A_u8_1' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_1[1];
                ^
test.c:935:17: note: previous definition is here
typedef uint8_t zig_A_u8_1[1];
                ^
test.c:993:17: error: redefinition of typedef 'zig_A_u8_4' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_4[4];
                ^
test.c:934:17: note: previous definition is here
typedef uint8_t zig_A_u8_4[4];
                ^
test.c:1019:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1052:17: error: redefinition of typedef 'zig_A_u8_3' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_3[3];
                ^
test.c:959:17: note: previous definition is here
typedef uint8_t zig_A_u8_3[3];
                ^
test.c:1106:17: error: redefinition of typedef 'zig_A_u8_5' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_5[5];
                ^
test.c:936:17: note: previous definition is here
typedef uint8_t zig_A_u8_5[5];
                ^
test.c:1198:17: error: redefinition of typedef 'zig_A_u8_10' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_10[10];
                ^
test.c:989:17: note: previous definition is here
typedef uint8_t zig_A_u8_10[10];
                ^
test.c:1281:17: error: redefinition of typedef 'zig_A_u8_8' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_8[8];
                ^
test.c:925:17: note: previous definition is here
typedef uint8_t zig_A_u8_8[8];
                ^
test.c:1317:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1332:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1352:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1367:17: error: redefinition of typedef 'zig_A_i32_5' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef int32_t zig_A_i32_5[5];
                ^
test.c:1114:17: note: previous definition is here
typedef int32_t zig_A_i32_5[5];
                ^
test.c:1370:17: error: redefinition of typedef 'zig_A_i32_5' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef int32_t zig_A_i32_5[5];
                ^
test.c:1367:17: note: previous definition is here
typedef int32_t zig_A_i32_5[5];
                ^
test.c:1379:17: error: redefinition of typedef 'zig_A_u8_6' is a C11 feature [-Werror,-Wtypedef-redefinition]
typedef uint8_t zig_A_u8_6[6];
                ^
test.c:1115:17: note: previous definition is here
typedef uint8_t zig_A_u8_6[6];
                ^
test.c:1458:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1517:9: error: empty struct is a GNU extension [-Werror,-Wgnu-empty-struct]
typedef struct {
        ^
test.c:1687:1: error: empty union is a GNU extension [-Werror,-Wgnu-empty-struct]
union {
^
test.c:1720:1: error: empty union is a GNU extension [-Werror,-Wgnu-empty-struct]
union {
^
test.c:1735:1: error: empty union is a GNU extension [-Werror,-Wgnu-empty-struct]
union {
^
test.c:1751:1: error: empty union is a GNU extension [-Werror,-Wgnu-empty-struct]
union {
^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

### LINKED ISSUE #12751: Naming mangling issue
### Zig Version

0.10.0-dev.3877+3deb33fff

### Steps to Reproduce

Create file.zig:
```zig
const std = @import("std");

pub fn main() void {
    const stdout = std.io.getStdOut().writer();
    stdout.print("Hello, {s}!\n", .{"world"}) catch unreachable;
}
```
Then run `zig build-exe -ofmt=c file.zig` and `zig cc -I/path/to/zig/lib/zig/include -w ./file.c`.

### Expected Behavior

Compilation completes without error.

### Actual Behavior

```
$ zig cc -I/path/to/zig/lib/zig/include -w ./file.c
./file.c:888:52: error: use of undeclared identifier 'closing'
     t13 = (zig_E_fmt_Parser_specifier ){ .error = zig_error_Expected closing ] };
                                                   ^
./file.c:7:28: note: expanded from macro 'zig_error_Expected'
#define zig_error_Expected closing ] 14
                           ^
./file.c:1225:61: error: use of undeclared identifier 'fmt_Parser_until__anon_3019'
   return (zig_L_u8 ){(uint8_t const * )((uint8_t const * )&fmt_Parser_until__anon_3019), 0u};
                                                            ^
2 errors generated.
```

### LINKED ISSUE #13068: stage2 crash with global ArrayList
### Zig Version

0.10.0-dev.4217+9d8cdb855

### Steps to Reproduce

```
const std = @import("std");

pub const allocator = std.heap.page_allocator;
var list = std.ArrayList(u32).init(allocator);

pub fn main() void {
    list.items.len = 0;
}
```

### Expected Behavior

Does not crash.

### Actual Behavior

Compiles fine with stage1, but not stage2:

```
$ zig build-exe main.zig 
Trace/breakpoint trap
```

Crash disappears if `list.items.len = 0` is replaced with `list.shrinkRetainingCapacity(0)`.

### PR DESCRIPTION
**Pull request by jacobly0**
This enabled all of the behavior tests for the C backend that were passing and fixed all of the C warnings from the flags `-Wextra -Wall -pedantic` for the behavior tests except various unused (but set) and uninitalized (only with safety off) warnings, and then fixes a bunch of disabled behavior tests.

Some notable C backend milestones achieved with these changes:
 * The C backend is now passing 90% (1243/1374) of the behavior tests vs the LLVM backend.
 * The official [hello world](https://ziglang.org/documentation/master/#Hello-World) now works on at least `x86_64-linux` (with and without libc).
 * The test runner now prints output to the console.

Fixes #12751
Fixes #11651

### CODE REVIEW

**Comment by Luukdegram** [comment]
I know this PR is still in draft, but could you extract 1cc6a95 into its own PR? It will make this and the new PR easier to review and we can merge the fixes separately from the C backend improvements. Especially as the fix contains changes to all of the native backends also, which was quite easy to miss from the title and description of this PR.

**Comment by andrewrk** [comment]
Looking forward to this!

**Comment by andrewrk** [comment]
> std: add cbe hacks to more targets 

Interesting :thinking: 

I'm going to look into enhancing the lowering of inline assembly, so that it can automatically do these transforms (e.g. turning "{x8}" to "{i}" and inserting a "mov". In the meantime, these hacks are nice workarounds :+1:

**Comment by andrewrk** [comment]
Related: #13278

**Comment by andrewrk** [comment]
Is this still intentionally a draft, or are you perhaps looking for review & merge now?

**Comment by jacobly0** [comment]
This is in as good a state as any to review and merge.  I can always open a new PR with further fixes.

File: lib/include/zig.h [hunk 1]
```diff
@@ -143,1457 +157,1347 @@
...
+#define    zig_atomic_load(obj,      order) zig_unimplemented()
 #define zig_fence(order) zig_unimplemented()
 #endif
 
-#include <stdint.h>
-#include <stddef.h>
-#include <limits.h>
+#if __STDC_VERSION__ >= 201112L
+#define zig_noreturn _Noreturn void
+#define zig_threadlocal thread_local
```
Comment (andrewrk) [inline lib/include/zig.h:179]: I see `#define zig_threadlocal` near the top of the file too

**Review by andrewrk — APPROVED** [review]
Incredible work. This is ready to be merged.

File: lib/include/zig.h (hunk 1 above)
Comment (jacobly0) [reply lib/include/zig.h]: Good catch,  looks like I got sidetracked in the middle of changing this.

**Pull request merged**